src/mqt/bench/benchmark_generator.py

```diff
@@ -3,429 +3,262 @@
-        pass
 
-    def timeout_handler(_signum, _frame):  # Custom signal handler
-        raise TimeoutException()
-
-    # Change the behavior of SIGALRM
-    signal.signal(signal.SIGALRM, timeout_handler)
-    signal.alarm(timeout)
-    try:
-        res = func(*args)
-    except TimeoutException:
-        print(
-            "Calculation/Generation exceeded timeout limit for ",
-            func.__name__,
-            func.__module__.split(".")[-1],
-            args[0].name,
-            args[1:],
-        )
-        return False
-    except Exception as e:
-        print(
-            "Exception: ",
-            e,
-            func.__name__,
-            func.__module__.split(".")[-1],
-            args[0].name,
-            args[1:],
-        )
-        return False
-    else:
-        # Reset the alarm
-        signal.alarm(0)
-
-    return res
 
+class BenchmarkGenerator:
+    def __init__(self, cfg_path: str = "./config.json", qasm_output_path: str | None = None):
+        with Path(cfg_path).open() as jsonfile:
+            self.cfg = json.load(jsonfile)
+            print("Read config successful")
+        self.timeout = self.cfg["timeout"]
+        if qasm_output_path is None:
+            self.qasm_output_path = str(resources.files("mqt.benchviewer") / "static" / "files" / "qasm_output")
+        else:
+            self.qasm_output_path = qasm_output_path
 
-def qc_creation_watcher(func, args):
-    class TimeoutException(Exception):  # Custom exception class
-        pass
+        Path(self.qasm_output_path).mkdir(exist_ok=True, parents=True)
 
-    def timeout_handler(_signum, _frame):  # Custom signal handler
-        raise TimeoutException
+    def create_benchmarks_from_config(self) -> bool:
+        Parallel(n_jobs=-1, verbose=100)(
+            delayed(self.generate_benchmark)(benchmark) for benchmark in self.cfg["benchmarks"]
+        )
+        return True
+
+    def generate_benchmark(self, benchmark):  # noqa: PLR0912, PLR0915
+        lib = utils.get_module_for_benchmark(benchmark["name"])
+        file_precheck = benchmark["precheck_possible"]
+        if benchmark["include"]:
+            if benchmark["name"] == "grover" or benchmark["name"] == "qwalk":
+                for anc_mode in benchmark["ancillary_mode"]:
+                    for n in range(
+                        benchmark["min_qubits"],
+                        benchmark["max_qubits"],
+                        benchmark["stepsize"],
+                    ):
+                        res_qc_creation = timeout_watcher(lib.create_circuit, self.timeout, [benchmark, n, anc_mode])
+                        if not res_qc_creation:
+                            break
+                        res = self.generate_circuits_on_all_levels(
+                            res_qc_creation, res_qc_creation.num_qubits, file_precheck
+                        )
+                        if not res:
+                            break
```

Comment:
This code fragment seems quite a bit redundant to me as it is repeated multiple times down below. Could be unified somehow.

Reply:
Done.